cpu/atmega_common/periph/i2c.c

```diff
@@ -0,0 +1,351 @@
+    /* enable device */
+    TWCR |= (1 << TWEN);
+
+    return 0;
+}
+
+int i2c_acquire(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+    mutex_lock(&lock);
+    return 0;
+}
+
+int i2c_release(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+    mutex_unlock(&lock);
+    return 0;
+}
+
+int i2c_read_byte(i2c_t dev, uint8_t address, void *data)
+{
+    return i2c_read_bytes(dev, address, data, 1);
+}
+
+int i2c_read_bytes(i2c_t dev, uint8_t address, void *data, int length)
+{
+    uint8_t *my_data = data;
+
+    assert((dev < I2C_NUMOF) && (length > 0));
+
+    /* send start condition and slave address */
+    if (_start(address, I2C_FLAG_READ) != 0) {
+        return 0;
+    }
+
+    for (int i = 0; i < length; i++) {
+        /* Send NACK for last received byte */
+        if ((length-i) == 1) {
+            TWCR = (1 << TWEN) | (1 << TWINT);
+        }
+        else {
+            TWCR = (1 << TWEA) | (1 << TWEN) | (1 << TWINT);
+        }
+        DEBUG("Wait for byte %i\n", i+1);
+        /* Wait for TWINT Flag set. This indicates that DATA has been received.*/
+        while (!(TWCR & (1 << TWINT))) {}
+        /* receive data byte */
+        my_data[i] = TWDR;
+        DEBUG("Byte %i received\n", i+1);
+    }
+
+    /* end transmission */
+    _stop();
+
+    return length;
+}
+
+int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, void *data)
+{
+    return i2c_read_regs(dev, address, reg, data, 1);
+}
+
+int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg, void *data, int length)
+{
+    assert((dev < I2C_NUMOF) && (length > 0));
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0) {
+        return 0;
+    }
+
+    /* send register address and wait for complete transfer to be finished*/
+    if (_write(&reg, 1) != 1) {
+        _stop();
+        return 0;
+    }
+
+    /* now start a new start condition and receive data */
+    return i2c_read_bytes(dev, address, data, length);
+}
+
+int i2c_write_byte(i2c_t dev, uint8_t address, uint8_t data)
+{
+    return i2c_write_bytes(dev, address, &data, 1);
+}
+
+int i2c_write_bytes(i2c_t dev, uint8_t address, const void *data, int length)
+{
+    int bytes = 0;
+
+    assert((dev < I2C_NUMOF) && (length > 0));
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0) {
+        return 0;
+    }
+
+    /* send out data bytes */
+    bytes = _write(data, length);
+
+    /* end transmission */
+    _stop();
+
+    return bytes;
+}
+
+int i2c_write_reg(i2c_t dev, uint8_t address, uint8_t reg, uint8_t data)
+{
+    return i2c_write_regs(dev, address, reg, &data, 1);
+}
+
+int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, const void *data, int length)
+{
+    int bytes = 0;
+
+    assert((dev < I2C_NUMOF) && (length > 0));
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0) {
+        return 0;
```

Comment:
true that. leave as is, might be addressed in a more general rework of i2c interface.